docs/DesignDocs/Protected_Code_Loader_Proposal.md

```diff
@@ -0,0 +1,161 @@
+Open Enclave Protected Code Loader
+====
+
+Enabling Enclave Code Confidentiality in Open Enclave.
+
+# Motivation
+Current Open Enclave SDK provides code integrity and data confidentiality at
+run-time but not enclave binary confidentiality on a disk. The enclave binary
+is in plaintext and can be reverse engineered to reveal code logic and
+secret data embedded in the enclave image. 
+
+The Open Enclave Protected Code Loader shall enable providing confidentiality 
+and integrity to the IP sections in enclave image, i.e. , the ELF file.
+
+# User Experience
+## Enclave Build-time:
+### Linking pcl lib to the enclave 
+On building the enclave, a static lib "liboepcl.a" should be linked to the 
+enclave image with the compilation option "-Wl, whole-archive"
+### Encrypt the enclave image
+An encryption tool is provided as the pre-step to enclave signing process:
+
+```bash
+oeencrypt -i my_enclave -o my_enclave.enc -k keyfile
+```
+
+## Enclave Load-time:
+On Enclave Loading, if the image is encrypted, it need to be decrypted 
+inside enclave loader, this step is done by OE SDK runtime and invisible 
+to user. 
+
+# Specification
+
+## Protected Code Loader Software Work Flow in Open Enclave
+### Build time encryption 
+    A seperate tool "oeencrypt" is provided at enclave build time to encrypt 
+	the enclave image right before enclave signing process.  
+### Load time decryption
+    On enclave initliazation operation(EINIT operation), right before relocation, 
+    the encrypted enclave image need to be decrypted and than perform the relocation
+    operation.
+
+## Encryption Algorithm in Protected Code Loader
+Using openssl AES-256-GCM as the encryption/decryption algorithm.
+
+## ELF sections Left plaintext
+### PCL table entry 
+A PCL table entry (a section called ".pcltbl" in ELF file)is built into the Enclave 
+image by linking to PCL lib, i.e. liboepcl.a, this part contains enclave decryption
+info on enclave loading and must remain plaintext.
+
+Definition of PCL entry(section ".pcltbl"):
+typedef struct pcl_table_t_
+{
+	/* Current state of PCL: initailized to PCL_PLAIN */
+    pcl_status_e pcl_state;                   
+    uint32_t     reserved1[3];                /* Must be 0 */
+	// GUID must match GUID in Sealed blob
+    uint8_t      pcl_guid[SGX_PCL_GUID_SIZE]; 
```

Comment:
This section is SGX-only but the notion of loading protected code is desired in other TEEs as well.   Can we cleanly separate TEE-agnostic functionality from TEE-specific functionality through some clean layer/interface?

Reply:
ok, I will consider it.

Reply:
pcl_guid is actually the aad for decryption key seal/unsealing functions. This is not SGX only, the MACRO SGX_PCL_GUID_SIZE is confusing here, I will change it in document. Also I will add more details on sealing/unsealing of the decryption key in this document.

Reply:
updated